linera-views/src/backends/lru_caching.rs

```diff
@@ -83,32 +128,149 @@ pub const DEFAULT_STORAGE_CACHE_CONFIG: StorageCacheConfig = StorageCacheConfig
     max_cache_size: 10000000,
     max_entry_size: 1000000,
     max_cache_entries: 1000,
+    max_cache_value_size: 10000000,
+    max_cache_find_keys_size: 10000000,
+    max_cache_find_key_values_size: 10000000,
 };
 
-enum CacheEntry {
+#[derive(Eq, Hash, PartialEq, Debug)]
+enum CacheKey {
+    Value(Vec<u8>),
+    FindKeys(Vec<u8>),
+    FindKeyValues(Vec<u8>),
+}
+
+enum ValueCacheEntry {
     DoesNotExist,
     Exists,
     Value(Vec<u8>),
 }
 
-impl CacheEntry {
+impl ValueCacheEntry {
     fn size(&self) -> usize {
         match self {
-            CacheEntry::Value(vec) => vec.len(),
+            ValueCacheEntry::Value(vec) => vec.len(),
             _ => 0,
         }
     }
 }
 
+struct FindKeysEntry(BTreeSet<Vec<u8>>);
+
+impl FindKeysEntry {
+    fn size(&self) -> usize {
+        self.0.iter().map(|key| key.len()).sum()
+    }
+
+    fn get_find_keys(&self, key_prefix: &[u8]) -> Vec<Vec<u8>> {
+        let key_prefix = key_prefix.to_vec();
+        let delta = key_prefix.len();
+        self.0
+            .range(get_interval(key_prefix))
+            .map(|key| key[delta..].to_vec())
+            .collect()
+    }
+
+    fn get_contains_key(&self, key: &[u8]) -> bool {
+        self.0.contains(key)
+    }
+
+    fn update_cache_entry(&mut self, key: &[u8], new_value: Option<Vec<u8>>) {
+        match new_value {
+            None => {
+                self.0.remove(key);
+            }
+            Some(_) => {
+                self.0.insert(key.to_vec());
+            }
+        }
+    }
+
+    fn delete_prefix(&mut self, key_prefix: &[u8]) {
+        let keys = self
+            .0
+            .range(get_interval(key_prefix.to_vec()))
+            .cloned()
+            .collect::<Vec<_>>();
+        for key in keys {
+            self.0.remove(&key);
+        }
+    }
+}
+
+struct FindKeyValuesEntry(BTreeMap<Vec<u8>, Vec<u8>>);
+
+impl FindKeyValuesEntry {
+    fn size(&self) -> usize {
+        self.0
+            .iter()
+            .map(|(key, value)| key.len() + value.len())
+            .sum()
+    }
+
+    fn get_find_keys(&self, key_prefix: &[u8]) -> Vec<Vec<u8>> {
+        let key_prefix = key_prefix.to_vec();
+        let delta = key_prefix.len();
+        self.0
+            .range(get_interval(key_prefix))
+            .map(|(key, _)| key[delta..].to_vec())
+            .collect()
+    }
+
+    fn get_find_key_values(&self, key_prefix: &[u8]) -> Vec<(Vec<u8>, Vec<u8>)> {
+        let key_prefix = key_prefix.to_vec();
+        let delta = key_prefix.len();
+        self.0
+            .range(get_interval(key_prefix))
+            .map(|(key, value)| (key[delta..].to_vec(), value.to_vec()))
+            .collect()
+    }
```

Comment:
If you returned `impl Iterator<Item = (Vec<u8>, Vec<u8>)>` then the caller could decide whether they want just the key, just value, or both. Both methods with one function.

Alternatively, it could be used internally – by these two – and expose the final API.